### PR TITLE
Use qos express

### PR DESCRIPTION
**Pull request by henrikstranneheim**
### Added

- Add express SLURM QOS
- Add new map to translate case priority from statusDB to SLURM QOS
- Add new function to get QOS from case priority
- Add tests

### Changed

- clincial trials QOS to "low"

### How to prepare for test
- [ ] ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] paxa the environment: `paxa`
- [ ] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh use_qos_express`

### How to test
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/workflow/analysis.py [hunk 1]
```diff
@@ -193,7 +194,7 @@ def add_pending_trailblazer_analysis(self, case_id: str) -> None:
             analysis_type=self.get_application_type(self.status_db.family(case_id).links[0].sample),
             out_dir=self.get_trailblazer_config_path(case_id=case_id).parent.as_posix(),
             config_path=self.get_trailblazer_config_path(case_id=case_id).as_posix(),
-            priority=self.get_priority_for_case(case_id=case_id),
+            priority=self.get_slurm_qos_for_case(case_id=case_id),
```
Comment (patrikgrenfeldt) [inline cg/meta/workflow/analysis.py:197]: great, in the best of worlds we should also change the variable `priority` to `qos`

File: cg/meta/workflow/balsamic.py [hunk 2]
```diff
@@ -469,7 +469,7 @@ def run_analysis(
             {
                 "--account": self.account,
                 "--mail-user": self.email,
-                "--qos": priority or self.get_priority_for_case(case_id=case_id),
+                "--qos": priority or self.get_slurm_qos_for_case(case_id=case_id),
```
Comment (patrikgrenfeldt) [inline cg/meta/workflow/balsamic.py:472]: same, we could also change the variable `priority` to `qos`

**Review by patrikgrenfeldt — APPROVED** [review]
Great improvement!

File: cg/meta/workflow/analysis.py (hunk 1 above)
Comment (henrikstranneheim) [reply cg/meta/workflow/analysis.py]: Done

File: cg/meta/workflow/balsamic.py (hunk 2 above)
Comment (henrikstranneheim) [reply cg/meta/workflow/balsamic.py]: Done

**Comment by henrikstranneheim** [comment]
@patrikgrenfeldt Do you have time for another review?

File: cg/cli/workflow/balsamic/base.py [hunk 3]
```diff
@@ -38,9 +38,9 @@
     default=False,
     help="Execute BALSAMIC in non-dry mode",
 )
-OPTION_PRIORITY = click.option(
-    "-p",
-    "--priority",
+OPTION_QOS = click.option(
+    "-qos",
+    "--slurm-quality-of-service",
     type=click.Choice([SlurmQos.LOW, SlurmQos.NORMAL, SlurmQos.HIGH]),
```
Comment (patrikgrenfeldt) [inline cg/cli/workflow/balsamic/base.py:44]: express?

File: cg/cli/workflow/mip/options.py [hunk 4]
```diff
@@ -3,9 +3,9 @@
 from cg.constants.priority import SlurmQos
 
 EMAIL_OPTION = click.option("-e", "--email", help="Email to send errors to", type=str)
-PRIORITY_OPTION = click.option(
-    "-p",
-    "--priority",
+QOS_OPTION = click.option(
+    "-qos",
+    "--slurm-quality-of-service",
     type=click.Choice([SlurmQos.LOW, SlurmQos.NORMAL, SlurmQos.HIGH]),
```
Comment (patrikgrenfeldt) [inline cg/cli/workflow/mip/options.py:9]: express?

File: cg/meta/workflow/analysis.py [hunk 5]
```diff
@@ -94,14 +94,15 @@ def all_flowcells_on_disk(self, case_id: str) -> bool:
...
-            return SlurmQos.LOW
-        if case_obj.priority_int > 1:
-            return SlurmQos.HIGH
-        return SlurmQos.NORMAL
+        return case_obj.priority.value or Priority.research
+
+    def get_slurm_qos_for_case(self, case_id: str) -> str:
+        """Get Quality of service (SLURM QOS) for the case"""
+        priority: int = self.get_priority_for_case(case_id)
+        return PRIORITY_TO_SLURM_QOS[priority]
```
Comment (patrikgrenfeldt) [inline cg/meta/workflow/analysis.py:105]: 👍

File: cg/meta/workflow/balsamic.py [hunk 6]
```diff
@@ -469,7 +469,7 @@ def run_analysis(
             {
                 "--account": self.account,
                 "--mail-user": self.email,
-                "--qos": priority or self.get_priority_for_case(case_id=case_id),
+                "--qos": slurm_quality_of_service or self.get_slurm_qos_for_case(case_id=case_id),
```
Comment (patrikgrenfeldt) [inline cg/meta/workflow/balsamic.py:472]: 💯

File: tests/cli/workflow/balsamic/test_run.py [hunk 7]
```diff
@@ -134,7 +134,7 @@ def test_priority_custom(cli_runner: CliRunner, balsamic_context: CGConfig, capl
     caplog.set_level(logging.INFO)
     # GIVEN valid case-id
     case_id = "balsamic_case_wgs_single"
-    priority_key = "--priority"
+    priority_key = "--slurm-quality-of-service"
```
Comment (patrikgrenfeldt) [inline tests/cli/workflow/balsamic/test_run.py:137]:
```suggestion
    qos_key = "--slurm-quality-of-service"
```

File: tests/cli/workflow/balsamic/test_run.py [hunk 8]
```diff
@@ -134,7 +134,7 @@ def test_priority_custom(cli_runner: CliRunner, balsamic_context: CGConfig, capl
...
     priority_value = SlurmQos.HIGH
```
Comment (patrikgrenfeldt) [inline tests/cli/workflow/balsamic/test_run.py:138]:
```suggestion
    qos_value = SlurmQos.HIGH
```

File: tests/meta/workflow/test_analysis.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    "priority,expected_slurm_qos",
+    [
+        (Priority.clinical_trials, SlurmQos.LOW),
+        (Priority.research, SlurmQos.LOW),
+        (Priority.standard, SlurmQos.NORMAL),
+        (Priority.priority, SlurmQos.HIGH),
+        (Priority.express, SlurmQos.EXPRESS),
+    ],
+)
+def test_get_priority_for_case(mocker, case_id: str, priority, expected_slurm_qos):
```
Comment (patrikgrenfeldt) [inline tests/meta/workflow/test_analysis.py:20]:
```suggestion
def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_qos):
```

**Review by patrikgrenfeldt — APPROVED** [review]
Great improvement to the overall codebase! 👍

**Review by barrystokman — APPROVED** [review]
SbatchDragen looks okay to me!

File: cg/cli/workflow/balsamic/base.py (hunk 3 above)
Comment (henrikstranneheim) [reply cg/cli/workflow/balsamic/base.py]: Good catch! Added

File: cg/cli/workflow/mip/options.py (hunk 4 above)
Comment (henrikstranneheim) [reply cg/cli/workflow/mip/options.py]: Fixed

File: cg/meta/workflow/balsamic.py (hunk 6 above)
Comment (henrikstranneheim) [reply cg/meta/workflow/balsamic.py:472]: 😄

File: tests/cli/workflow/balsamic/test_run.py (hunk 7 above)
Comment (henrikstranneheim) [reply tests/cli/workflow/balsamic/test_run.py]: Fixed

File: tests/cli/workflow/balsamic/test_run.py (hunk 8 above)
Comment (henrikstranneheim) [reply tests/cli/workflow/balsamic/test_run.py]: fixed

File: tests/meta/workflow/test_analysis.py (hunk 9 above)
Comment (henrikstranneheim) [reply tests/meta/workflow/test_analysis.py]: Good catch! Fixed

**Review by Mropat — APPROVED** [review]

**Comment by henrikstranneheim** [comment]
@Vince-janv MIP is ready: https://github.com/Clinical-Genomics/MIP/pull/1979

**Comment by Vince-janv** [comment]
`cg deliver rsync` works as intended. SLURM script create in stage runs and completes when submitted in prod environment:

Screenshot 2022-03-16 at 14 04 21
Screenshot 2022-03-16 at 14 05 48

**Comment by Vince-janv** [comment]
SLURM scripts generated by `cg demultiplex flowcell` in stage and prod are identical:

Screenshot 2022-03-16 at 14 34 47

**Comment by Vince-janv** [comment]
With the changes made in https://github.com/Clinical-Genomics/trailblazer/pull/179, MIP-dna and Balsamic analyses with qos express can now be run by the cg CLI.

**Comment by henrikstranneheim** [comment]
@Vince-janv Awesome! Then we are ready to merge next week! Well done!

**Pull request merged**